rfd/0147-repo-mirror.md

```diff
@@ -0,0 +1,73 @@
+* `sync/upstream-[UPSTREAM_BRANCH_NAME]` - This branch will be an exact copy of the upstream branch. Should there be a need to create a PR in this repository, the history should be based on this branch, not on `master`.
+* `sync/rebase` - This is the branch where custom changes will be committed. It will be rebased onto the upstream changes before being merged into master.
+
+#### Branch Protections
+
+The only protected branch will be `sync/rebase`. Other branches will allow force pushes to enable forceful synchronization based on the upstream state.
+
+### High Level Implementation
+
+The implementation has two primary components:
+1. A GitHub Action, scheduled and also triggered by changes to the `sync/rebase` branch. This workflow will handle the rebase operation and commit the results to the appropriate branches (described above).
+2. A script invoked when the rebase cannot be applied cleanly. This will enable custom rebase logic, reducing the frequency with which the `sync/rebase` branch needs manual updates to avoid rebase conflicts. A simple example is that workflow automations not desired in the mirror repository might be removed in the `sync/rebase` branch. Any modifications to these workflows would then cause a conflict that can be resolved by simply removing the files. This logic can be changed to meet the specific needs of the repository.
+
+At a high level the action will execute the following steps:
+1. Set up authentication and any repository-specific configurations.
+2. Check out the latest upstream contents.
+3. Fetch our custom changes.
+4. Push the exact copy branches `sync/upstream-[UPSTREAM_BRANCH_NAME]` to our mirror.
+5. Rebase our `sync/rebase` branch onto the latest upstream changes, using the script for conflict resolutions when necessary and possible.
+6. Force-push the rebased changes to the default branch of our repository mirror
+
+### Specific Implementation Details
+
+#### teleport-private
+
+The only custom changes required for `teleport-private` is the need to maintain multiple `sync/upstream` branches. Specifically, we will sync all release branches. This will be useful when producing backport PRs. Despite syncing multiple upstream branches, only the `master` branch will undergo rebasing (as generally documented above). The rebase is necessary only to incorporate changes related to GitHub Actions and Workflow automations required for syncing.
+
+#### Teleport Security Scanning
+
+This tooling will leveraged to address gaps in our release branch scanning in the following manner:
+1. Create four repositories, one for each supported version (and one extra described below): `teleport-sec_scan-1`, `teleport-sec_scan-2`, `teleport-sec_scan-3`, and `teleport-sec_scan-4`. The 1 will refer to our latest release branch, determined by automation, with each subsequent repository covering a consecutively older version.
```

Comment:
Seems a bit confusing to not have them mapped to release branch identifiers. How do we know what the mapping means?

Reply:
The reason for not including the release branch identifier in the repo name is to avoid having a repo per release version.  This will allow a constant set of repos vs needing manual action to create and configure a new repo for every version we release.

Right now you would need to check the workflow logs for what version is being targeted.

Reply:
I could see this getting messy very quickly. I get not wanting to create a new repository, but we already do a lot of manual work with every new major version, so not sure it's really worth it. Anything that obfuscates engineers from understanding what version things map to is likely to cause confusion. But we can try it and see how it goes.

Reply:
We can see, I don't think it will be too bad since the order is consistent.
1 = Always the latest release branch
2 = The prior release
... and so on

I expect these repos to also primarily be used by us in the security team initially.  We will likely need more automation in the future for getting alerts out of these repos and into the upstream `teleport` repo where people will notice them.

Reply:
I have updated the naming to only use `-` characters.

I still am not convinced that having a repo for every version makes sense.  There was [some concern earlier](https://github.com/gravitational/teleport/pull/32233#discussion_r1332092447) about the quantity of repos that would result with.  And I dislike any manual process needed here.  Beyond just the manual effort it also creates a potential source of error in our security reporting (which would then potentially need monitoring to prevent).

Having this be done with some form of automation overall makes this simpler.  That said, perhaps this is not the best form of it?  Let me know if anyone has any further suggestions